src/core/geometry/qgsgeometryutils.h

```diff
@@ -682,6 +682,16 @@ class CORE_EXPORT QgsGeometryUtils
       return QgsGeometryUtilsBase::sqrDistance2D( pt1.x(), pt1.y(), pt2.x(), pt2.y() );
     }
 
+    /**
+     * Returns the squared 3D distance between two points.
+     *
+     * No check is done, if z contains NaN value.
+     */
+    static double sqrDistance3D( const QgsPoint &pt1, const QgsPoint &pt2 ) SIP_HOLDGIL
+    {
+      return QgsGeometryUtilsBase::sqrDistance3D( pt1.x(), pt1.y(), pt1.z(), pt2.x(), pt2.y(), pt2.z() );
```

Comment:
You should add this check:

```cpp
 if ( is3D() || !std::isnan( z ) )
    zDistSquared = ( mZ - z ) * ( mZ - z );
```

Reply:
No, it's intentional, and it's precisely indicated in the docstring.

We've often had this discussion, especially with @m-kuhn, to figure out what the hell we're doing with these NaNs.
Here, it's a basic method; we don't do the check. It's done upstream, for QgsPoint for example: https://github.com/qgis/QGIS/pull/55619/files#diff-0985eb68b901e2caee7ea8b773bc894f34898a0fde1b352e2a89cae8a8d84f80R424. Here, to stay consistent with the logic introduced by Matthias in https://github.com/qgis/QGIS/pull/4712.